### ISSUE TITLE
Autoreconnect issue with ESP32

### ISSUE THREAD

**Issue by steros76** [plapointe6/EspMQTTClient#33]
If ESP lose Wifi signal, this lib should try to reconnect. Now connection is lost until I reset ESP.
I think this is very important for IOT devices.

**Comment by plapointe6** [comment]
Thanks for reporting this, I will make this issue my priority.

**Comment by abqmichael** [comment]
Hi. I just pinned down this issue myself when running ESP32 boards. I had noticed that my device was going zombie randomly. In tracking this down, I programmed 3 different ESP32 boards to run the same software and found that they all went zombie together. Looking on my network, I found that my router was misbehaving (a hard reset cured that). Moving back to the tests, I found that the clients would reconnect just fine if I unplugged my MQTT server for several minutes but do not appear to reliably reconnect if I unplugged my router. (I only tested that once.)

In parallel to the tests, I found an article describing the same problem reported related to PubSubClient, an underlying libraries used in EspMQTTClient. In their investigation, they found that the problem was more with the example code than PubSubClient itself. Apparently the connection code needed to test a variety of conditions. They concluded that software should connect very differently than in the example and they gave some example code for doing this. (https://github.com/knolleary/pubsubclient/issues/444).

I only barely started investigating whether this would be right for EspMQTTClient. Within the code, I see that the connect/reconnect software is written quite differently than the discussion. As a result, it might not be as simple to patch as I hoped. However, their guide may be useful just the same.

**Comment by kkberg** [comment]
Hello, having also connection issue when collecting data only every 15 minutes and want to push to MQTT-broker. I then tried to use a loop instead of only delay() function. But after 2 connects in 2 minutes connection ist lost.
![mqtt](https://user-images.githubusercontent.com/1121884/69866181-bb537700-12a3-11ea-8151-68fa277d0e97.JPG)
Thank you

**Comment by plapointe6** [comment]
Have-you a ESP8266 or ESP32 ? I am pretty sure that the connection issue arise with ESP32 only. It will guide my troubleshooting.

**Comment by kkberg** [comment]
I have an ESP8266 12E (WEMOS D1 Mini) it also happens on other NodeMCU's like the Lolin .

**Comment by plapointe6** [comment]
@kkberg make sure that you never "block" the main core for too long, as it will result to a connection timeout. The reason is that there is only one core to handle the communication and to execute your code. Don't use the delay function. Exemple of what you could do : 
 ```c++

#include "EspMQTTClient.h"

EspMQTTClient client(
.....
);

void setup()
{
.....
}

void onConnectionEstablished()
{
  // Execute delayed instructions, here is a 5 seconds delay
  client.executeDelayed(5 * 1000, readSensor);
}

void readSensor() 
{
	// IMPORTANT, protect from spawning multiple delayed instructions when the connection is lost
	if(!client.isConnected())
		return

	// What you wanbt to execute each (5) seconds
	client.publish("mytopic/test", "This is a message sent with a delay of 5 seconds");
	
	// Re-schedule the instructions
	client.executeDelayed(5 * 1000, readSensor);
}

void loop()
{
  client.loop();
}
```

**Comment by kkberg** [comment]
@plapointe6 

Thank you very much, I integrated your solution and I got it to work.

**Comment by MoosHam** [comment]
Hi all,

I know the failing reconnect on lost Wifi is a bug and will be dealt with. But how do you guys work around this right now? I also suffer from it but don't know how to deal with it, other than jsut occasionally rebooting the ESP32 once I suspect it is hanging...

thanks,
MoosHam

**Comment by abqmichael** [comment]
Sadly, I don't have a work-around. I have traced the problem to loosing WiFi. In fact, I can get it to reliably fail by just turning my WiFi router off and waiting a little.

I tried the suggestion about not sending messages during a callback and it didn't change anything for me. Frankly, it turned out to not be my problem. I decided to dig into the library code to see what I could find out. What I found is that after the WiFi gets lost, eventually EspMQTTClient::connectToMqttBroker() gets called (probably from from EspMQTTClient::loop()) and this, in turn, calls mMqttClient.connect (mMqttClient is a PubSubClient instance) and that method hangs.
 
Tracing down from EspMQTTClient::loop(), I see that connectToMqttBroker() gets called when loop() discovers that it has lost the MQTT client connection. Oddly, though loop() does discover that, it does not recognize at that time that it has lost WiFi. That is, it appears that WiFi.status() == WL_CONNECTED even though the WiFi is gone. 

Right now, I'm trying to see if I can fake things out so that the ESP32 forces a disconnect when it looses WiFi and then will re-establish then. However, my first stab at that idea failed so perhaps I'll have to dig further.

As an aside, I've read that other people are having problems keeping their ESP32s connected to specific routers. My router (an Apple Time capsule) is on that list of bad boys. I might punt and buy a different wifi router.

Michael
> 
> I know the failing reconnect on lost Wifi is a bug and will be dealt with. But how do you guys work around this right now? I also suffer from it but don't know how to deal with it, other than jsut occasionally rebooting the ESP32 once I suspect it is hanging...
> 
> thanks,
> MoosHam

**Comment by MoosHam** [comment]
Hi Michael,

great to see you seem to be getting to the bottom of this issue. Quite annoying little bug. BTW, I use a ubiquity AP and have similiar issues. It's happening only from time to time (every few days or so), which makes it hard to catch...

MoosHam

**Comment by abqmichael** [comment]
I’m getting confident that this can be done. I found some complicated code that used another MQTT library at https://esp32.com/viewtopic.php?t=3851 <https://esp32.com/viewtopic.php?t=3851>  I say complicated mainly because it also used the secure connection to MQTT and I haven’t gotten there in configuring my MQTT server. I modified this code and got it to work with my MQTT server and it survived the wifi re-boot.

I included my version below. Now I’m working to integrate the idea into the EspMQTTClient library. 

Michael

/* standalone device with MQTT function
    Code found at https://esp32.com/viewtopic.php?t=3851

   The requirements are as follows:
   - it should run its main tasks regardless of the availability of WiFi/MQTT
   - it should recover lost WiFi/MQTT without interrupting the main tasks
   - unless other examples on the net it does not start the WiFi/MQTT connection in the setup() section
     since the device should start with its main functions immediately and do the connection setup later

   Other features
   - OTA over WiFi using the Arduino IDE
   - MQTT over TLS for encrypted data transfer
   - configuration parameters for device name, SSID, WAP-PW, MQTT broker username and password are stored in Preferences

    Feedback to improve this code is welcome
    Urs Eppenberger
*/

#include <WiFi.h>
#include <MQTT.h>
// Shows up in Arduino Library as MQTT by Joel Gaehwiler

const char* Hostname = "esp32_wt";            // change according your setup : it is used in OTA and as MQTT identifier
String WiFi_SSID = "AbqMichaelJ";               // change according your setup : SSID and password for the WiFi network
String WiFi_PW = "BananaNutBread";                   //    "
const char* OTA_PW = "BananaNutBread";                // change according your setup : password for 'over the air sw update'
String mqtt_broker = "10.0.1.201";         // change according your setup : IP Adress or FQDN of your MQTT broker
String mqtt_user = "esp32_wt";               // change according your setup : username and password for authenticated broker access
String mqtt_pw = "";                   //    "
String input_topic = "wifi/test";        // change according your setup : MQTT topic for messages from device to broker
unsigned long waitCount = 0;                 // counter
uint8_t conn_stat = 0;                       // Connection status for WiFi and MQTT:
//
// status |   WiFi   |    MQTT
// -------+----------+------------
//      0 |   down   |    down
//      1 | starting |    down
//      2 |    up    |    down
//      3 |    up    |  starting
//      4 |    up    | finalising
//      5 |    up    |     up

unsigned long lastStatus = 0;                // counter in example code for conn_stat == 5
unsigned long lastTask = 0;                  // counter in example code for conn_stat <> 5

const char* Version = "{\"Version\":\"low_prio_wifi_v2\"}";
const char* Status = "{\"Message\":\"up and running\"}";

WiFiClient TCP;                        // TCP client object, uses SSL/TLS
MQTTClient mqttClient(512);                  // MQTT client object with a buffer size of 512 (depends on your message size)


void setup() {
  Serial.begin(115200);
  // configure mqttClient
}

String message = "None";

void loop() {                                                     // with current code runs roughly 400 times per second
  // start of non-blocking connection setup section
  if ((WiFi.status() != WL_CONNECTED) && (conn_stat != 1)) {
    conn_stat = 0;
  }
  if ((WiFi.status() == WL_CONNECTED) && !mqttClient.connected() && (conn_stat != 3))  {
    conn_stat = 2;
  }
  if ((WiFi.status() == WL_CONNECTED) && mqttClient.connected() && (conn_stat != 5)) {
    conn_stat = 4;
  }
  switch (conn_stat) {
    case 0:                                                       // MQTT and WiFi down: start WiFi
      Serial.println("MQTT and WiFi down: start WiFi");
      WiFi.begin(WiFi_SSID.c_str(), WiFi_PW.c_str());
      conn_stat = 1;
      break;
    case 1:                                                       // WiFi starting, do nothing here
      message = "WiFi starting ";
      if (waitCount < 5 || (waitCount % 100) == 0) {
        Serial.printf("Wifi starting %d\n", waitCount);
      }
      waitCount++;
      break;
    case 2:                                                       // WiFi up, MQTT down: start MQTT
      Serial.printf("WiFi up, MQTT down, waited %d: starting MQTT\n", waitCount);
      message = "WiFi up, MQTT down ";
      mqttClient.begin(mqtt_broker.c_str(), TCP);           //   config MQTT Server, use port 8883 for secure connection
      conn_stat = 3;
      waitCount = 0;
      break;
    case 3:                                                       // WiFi up, MQTT starting, do nothing here
      if (waitCount < 5 || (waitCount % 100) == 0) {
        Serial.printf("WiFi up, MQTT starting %d\n", waitCount);
      }
      message = "WiFi up, MQTT starting ";

      //if(mqttClient.connect(Hostname, mqtt_user.c_str(), mqtt_pw.c_str())) {
      //  conn_stat = 4;
      //}
      if(mqttClient.connect("arduino", "try", "try")) {
        Serial.println("\nconnected!");
      } else {
        Serial.print(".");
        delay(10);
      }      
      waitCount++;
      break;
    case 4:                                                       // WiFi up, MQTT up: finish MQTT configuration
      Serial.printf("WiFi up, MQTT up, waited %d: finish MQTT configuration\n", waitCount);
      message = "WiFi up, MQTT up ";
      //mqttClient.subscribe(output_topic);
      mqttClient.publish(input_topic, Version);
      conn_stat = 5;
      break;
  }
  // end of non-blocking connection setup section

  // start section with tasks where WiFi/MQTT is required
  if (conn_stat == 5) {
    if (millis() - lastStatus > 10000) {                            // Start send status every 10 sec (just as an example)
      Serial.println(Status);
      mqttClient.publish(input_topic, Status);                      //      send status to broker
      mqttClient.loop();                                            //      give control to MQTT to send message to broker
      lastStatus = millis();                                        //      remember time of last sent status message
    }
    mqttClient.loop();                                              // internal household function for MQTT
  }
  // end of section for tasks where WiFi/MQTT are required

  // start section for tasks which should run regardless of WiFi/MQTT
  if (millis() - lastTask > 1000) {                                 // Print message every second (just as an example)
    Serial.println(message);
    lastTask = millis();
  }
  delay(100);
  // end of section for tasks which should run regardless of WiFi/MQTT
}

**Comment by abqmichael** [comment]
As to work-around, the first thing that helped was to hard reset my apple time capsule and bring my network down to just that. (I had some apple devices extending my network and I turned them off.) I'm guessing there must have been a problem as the disconnect frequency went way down. You might wish to have a look at your wifi access point.

Oh... I got my test code working with PubSubClient so now it's time to start modifying the EspMQTTClient library. I'm hoping I can do that this week. I will only be able to post that as a suggested fix and Patrick will need to analyze/test and accept or reject it.

**Comment by plapointe6** [comment]
Great, I can help if you have questions. One of the challenge of this change will be to keep the ESP8266 compatibility.

**Comment by abqmichael** [comment]
Thankfully the code I'm patching doesn't have any of the ESP8266 specific calls in it. You had nicely separated that into methods called by the patch code. I got that working but I am having a problem right now with the callback execution. I may have inadvertently broken something. So I'm going to go back to your latest version and more carefully push in my patch with attention to keeping the code around that piece. 

One question I have is whether the callbacks need the MQTT connected. I think that's how it was implemented. However, it would be nice if the logic in my applications could keep running whether the network was up or down. (My application would run fine off network. MQTT is only needed to tell it to do new things.)

**Comment by plapointe6** [comment]
Not sure to understand your question, maybe some code can help. The callbacks are here to trigger when something is received from MQTT, so they won't be triggered if the MQTT is not connected. Maybe the problem you face is linked to #21, when the networks go down, the PubSubClient library blocks the sketch. It's an issue that I plan to resolve some day by implementing my own version of PubSubClient. Sometime there is workarounds to prevent that from happening, it depends on the cases.

**Comment by abqmichael** [comment]
I am referring to executeDelayed. One of my favorite features of EspMQTTClient is the whole event system vis executeDelayed. For example, I use it to animate lights vis FastLED. I want my light animation party to go on when the WiFi goes down. Besides not going back online, that’s what happens in the current release.

**Comment by abqmichael** [comment]
I have a version that's working and am now doing cleanup and optimization. How would you like me to get it to you?

**Comment by plapointe6** [comment]
If you plan to do a pull request, it's better if you finish your cleanup and create a pull request. This way I will easily be able to see changes and comment parts of the code :)

**Comment by abqmichael** [comment]
Hi Patrick;

I’m very novice at git and I don’t imagine that I have write access to your repository. (My brain just can’t parse git lingo. For example, I’ve no real idea what it means to "Generate a request asking your upstream project to pull changes into their tree.”)

If you will email me at abqmichaelj@aol.com <mailto:abqmichaelj@aol.com>, I can send you a copy of the code. Alternatively, I could put it in one of these comments. I suppose I could also do the git thing if you wish to make me a contributor to your project. 

To make my development and your seeing what I’ve done easier, I have demarked all my relevant changes vs the original between #ifdef WIFI_FIX/#else/#endif blocks. (This let me quickly go back and forth between versions.) I am working from the most recent version of the code that’s on GitHub.

**Comment by plapointe6** [comment]
To do a pull request and become a contributor of this project, make a fork of this lib, commit your changes to your fork and then you will be able to do a pull request : https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request-from-a-fork

If you prefer to not do this procedure, I can get your code and adapt it by my own.

**Comment by abqmichael** [comment]
#ifndef ESP_MQTT_CLIENT_H
#define ESP_MQTT_CLIENT_H

#include <PubSubClient.h>

#ifdef ESP8266

  #include <ESP8266WiFi.h>
  #include <ESP8266WebServer.h>
  #include <ESP8266mDNS.h>
  #include <ESP8266HTTPUpdateServer.h>
  
  #define WebServer ESP8266WebServer
  #define ESPmDNS ESP8266mDNS
  #define ESPHTTPUpdateServer ESP8266HTTPUpdateServer

#else // for ESP32

  #include <WiFiClient.h>
  #include <WebServer.h>
  #include <ESPmDNS.h>
  #include "ESP32HTTPUpdateServer.h"

  #define ESPHTTPUpdateServer ESP32HTTPUpdateServer

#endif

#define MAX_TOPIC_SUBSCRIPTION_LIST_SIZE 10
#define MAX_DELAYED_EXECUTION_LIST_SIZE 10
#define CONNECTION_RETRY_DELAY 10 * 1000

#define WIFI_FIX
#ifdef WIFI_FIX
#pragma message "Using NEW connection algorithm V 1/23/20 @ 10:45"
#endif

void onConnectionEstablished(); // MUST be implemented in your sketch. Called once everythings is connected (Wifi, mqtt).

typedef void(*ConnectionEstablishedCallback) ();
typedef void(*MessageReceivedCallback) (const String &message);
typedef void(*MessageReceivedCallbackWithTopic) (const String &topicStr, const String &message);
typedef void(*DelayedExecutionCallback) ();

class EspMQTTClient 
{
private:
#ifdef WIFI_FIX
  // mConnState values
  // status |   WiFi   |    MQTT
  // -------+----------+------------
  //      0 |   down   |    down
  //      1 | starting |    down
  //      2 |    up    |    down
  //      3 |    up    |  starting
  //      4 |    up    | finalising
  //      5 |    up    |     up
  uint8_t mConnState; // connection state (1-5)
  unsigned int mWaitCount;
  unsigned int mBrokerConnectPauseMills; // minimum time (mills) allowed between calls to mMqttClient.connect
  // Wifi related
#else
  bool mWifiConnected;
  unsigned long mLastWifiConnectionAttemptMillis;
  unsigned long mLastWifiConnectionSuccessMillis;
#endif
  const char* mWifiSsid;
  const char* mWifiPassword;
  WiFiClient mWifiClient;

  // MQTT related
  bool mMqttConnected;
  unsigned long mLastMqttConnectionMillis;
  const char* mMqttServerIp;
  const char* mMqttUsername;
  const char* mMqttPassword;
  const char* mMqttClientName;
  const short mMqttServerPort;
  bool mMqttCleanSession;
  char* mMqttLastWillTopic;
  char* mMqttLastWillMessage;
  bool mMqttLastWillRetain;

  PubSubClient mMqttClient;

  struct TopicSubscriptionRecord {
    String topic;
    MessageReceivedCallback callback;
    MessageReceivedCallbackWithTopic callbackWithTopic;
  };
  TopicSubscriptionRecord mTopicSubscriptionList[MAX_TOPIC_SUBSCRIPTION_LIST_SIZE];
  byte mTopicSubscriptionListSize;

  // HTTP update server related
  char* mUpdateServerAddress;
  char* mUpdateServerUsername;
  char* mUpdateServerPassword;
  WebServer* mHttpServer;
  ESPHTTPUpdateServer* mHttpUpdater;

  // Delayed execution related
  struct DelayedExecutionRecord {
    unsigned long targetMillis;
    DelayedExecutionCallback callback;
  };
  DelayedExecutionRecord mDelayedExecutionList[MAX_DELAYED_EXECUTION_LIST_SIZE];
  byte mDelayedExecutionListSize;

  // General behaviour related
  ConnectionEstablishedCallback mConnectionEstablishedCallback;
  bool mEnableSerialLogs;
  bool mShowLegacyConstructorWarning;
  unsigned int mConnectionEstablishedCount; // Incremented before each mConnectionEstablishedCallback call

public:
  // Wifi + MQTT with no MQTT authentification
  EspMQTTClient(
    const char* wifiSsid, 
    const char* wifiPassword, 
    const char* mqttServerIp, 
    const char* mqttClientName = "ESP8266",
    const short mqttServerPort = 1883);

  // Wifi + MQTT with MQTT authentification
  EspMQTTClient(
    const char* wifiSsid,
    const char* wifiPassword,
    const char* mqttServerIp,
    const char* mqttUsername,
    const char* mqttPassword,
    const char* mqttClientName = "ESP8266",
    const short mqttServerPort = 1883);

  // Only MQTT handling (no wifi), with MQTT authentification
  EspMQTTClient(
    const char* mqttServerIp,
    const short mqttServerPort,
    const char* mqttUsername,
    const char* mqttPassword,
    const char* mqttClientName = "ESP8266");

  // Only MQTT handling without MQTT authentification
  EspMQTTClient(
    const char* mqttServerIp,
    const short mqttServerPort,
    const char* mqttClientName = "ESP8266");

  // Legacy constructor for version 1.3 - WILL BE DELETED SOONER OR LATER
  EspMQTTClient(
    const char* wifiSsid, 
    const char* wifiPassword,
    ConnectionEstablishedCallback connectionEstablishedCallback,
    const char* mqttServerIp, 
    const short mqttServerPort = 1883,
    const char* mqttUsername = NULL, 
    const char* mqttPassword = NULL, 
    const char* mqttClientName = "ESP8266",
    const bool enableWebUpdater = true, 
    const bool enableSerialLogs = true);

  // Legacy constructor for version <= 1.2 - WILL BE DELETED SOONER OR LATER
  EspMQTTClient(
    const char* wifiSsid,
    const char* wifiPassword, 
    const char* mqttServerIp,
    const short mqttServerPort, 
    const char* mqttUsername, 
    const char* mqttPassword,
    const char* mqttClientName, 
    ConnectionEstablishedCallback connectionEstablishedCallback,
    const bool enableWebUpdater = true, 
    const bool enableSerialLogs = true);

  ~EspMQTTClient();

  // Optional functionality
  void enableDebuggingMessages(const bool enabled = true); // Allow to display useful debugging messages. Can be set to false to disable them during program execution
#ifdef WIFI_FIX
  void brokerConnectPause(unsigned int mills); // set minimum time (mills) allowed between calls to mMqttClient.connect 
#endif
  void enableHTTPWebUpdater(const char* username, const char* password, const char* address = "/"); // Activate the web updater, must be set before the first loop() call.
  void enableHTTPWebUpdater(const char* address = "/"); // Will set user and password equal to mMqttUsername and mMqttPassword
  void enableMQTTPersistence(); // Tell the broker to establish a persistent connection. Disabled by default. Must be called before the first loop() execution
  void enableLastWillMessage(const char* topic, const char* message, const bool retain = false); // Must be set before the first loop() call.

  // Main loop, to call at each sketch loop()
  void loop();

  // MQTT related
  bool publish(const String &topic, const String &payload, bool retain = false);
  bool subscribe(const String &topic, MessageReceivedCallback messageReceivedCallback);
  bool subscribe(const String &topic, MessageReceivedCallbackWithTopic messageReceivedCallback);
  bool unsubscribe(const String &topic);   //Unsubscribes from the topic, if it exists, and removes it from the CallbackList.

  // Other
  void executeDelayed(const unsigned long delay, DelayedExecutionCallback callback);
#ifdef WIFI_FIX
  bool isConnected(); // Return true if everything is connected
  bool isWifiConnected() ; // Return true if wifi is connected
  bool isMqttConnected(); // Return true if mqtt is connected
#else
  inline bool isConnected() const { return isWifiConnected() && isMqttConnected(); }; // Return true if everything is connected
  inline bool isWifiConnected() const { return mWifiConnected; }; // Return true if wifi is connected
  inline bool isMqttConnected() const { return mMqttConnected; }; // Return true if mqtt is connected
#endif
  inline bool getConnectionEstablishedCount() const { return mConnectionEstablishedCount; }; // Return the number of time onConnectionEstablished has been called since the beginning.

  inline void setOnConnectionEstablishedCallback(ConnectionEstablishedCallback callback) { mConnectionEstablishedCallback = callback; }; // Default to onConnectionEstablished, you might want to override this for special cases like two MQTT connections in the same sketch

private:
  void connectToWifi();
  void connectToMqttBroker();
  bool mqttTopicMatch(const String &topic1, const String &topic2);
  void mqttMessageReceivedCallback(char* topic, byte* payload, unsigned int length);
#ifdef WIFI_FIX
  void configureHTTPWebUpdater(); // Worker method to configure web updater. 
#endif
};

#endif

**Comment by abqmichael** [comment]
#include "EspMQTTClient.h"


// =============== Constructor / destructor ===================

// MQTT only (no wifi connection attempt)
EspMQTTClient::EspMQTTClient(
  const char* mqttServerIp,
  const short mqttServerPort,
  const char* mqttClientName) :
  EspMQTTClient(NULL, NULL, mqttServerIp, NULL, NULL, mqttClientName, mqttServerPort)
{
}
EspMQTTClient::EspMQTTClient(
  const char* mqttServerIp,
  const short mqttServerPort,
  const char* mqttUsername,
  const char* mqttPassword,
  const char* mqttClientName) :
  EspMQTTClient(NULL, NULL, mqttServerIp, mqttUsername, mqttPassword, mqttClientName, mqttServerPort)
{
}

// Wifi and MQTT handling
EspMQTTClient::EspMQTTClient(
  const char* wifiSsid,
  const char* wifiPassword,
  const char* mqttServerIp,
  const char* mqttClientName,
  const short mqttServerPort) :
  EspMQTTClient(wifiSsid, wifiPassword, mqttServerIp, NULL, NULL, mqttClientName, mqttServerPort)
{
}

// Warning : for old constructor support, this will be deleted soon or later
EspMQTTClient::EspMQTTClient(
  const char* wifiSsid, 
  const char* wifiPassword, 
  const char* mqttServerIp,
  const short mqttServerPort, 
  const char* mqttUsername, 
  const char* mqttPassword,
  const char* mqttClientName, 
  ConnectionEstablishedCallback connectionEstablishedCallback,
  const bool enableWebUpdater, 
  const bool enableSerialLogs) :
  EspMQTTClient(wifiSsid, wifiPassword, mqttServerIp, mqttUsername, mqttPassword, mqttClientName, mqttServerPort)
{
  if (enableWebUpdater)
    enableHTTPWebUpdater();

  if (enableSerialLogs)
    enableDebuggingMessages();

  setOnConnectionEstablishedCallback(connectionEstablishedCallback);

  mShowLegacyConstructorWarning = true;
}

// Warning : for old constructor support, this will be deleted soon or later
EspMQTTClient::EspMQTTClient(
  const char* wifiSsid, 
  const char* wifiPassword,
  ConnectionEstablishedCallback connectionEstablishedCallback, 
  const char* mqttServerIp, 
  const short mqttServerPort,
  const char* mqttUsername,
  const char* mqttPassword, 
  const char* mqttClientName,
  const bool enableWebUpdater,
  const bool enableSerialLogs) :
  EspMQTTClient(wifiSsid, wifiPassword, mqttServerIp, mqttUsername, mqttPassword, mqttClientName, mqttServerPort)
{
  if (enableWebUpdater)
    enableHTTPWebUpdater();

  if (enableSerialLogs)
    enableDebuggingMessages();

  setOnConnectionEstablishedCallback(connectionEstablishedCallback);

  mShowLegacyConstructorWarning = true;
}

EspMQTTClient::EspMQTTClient(
  const char* wifiSsid,
  const char* wifiPassword,
  const char* mqttServerIp,
  const char* mqttUsername,
  const char* mqttPassword,
  const char* mqttClientName,
  const short mqttServerPort) :
  mWifiSsid(wifiSsid),
  mWifiPassword(wifiPassword),
  mMqttServerIp(mqttServerIp),
  mMqttUsername(mqttUsername),
  mMqttPassword(mqttPassword),
  mMqttClientName(mqttClientName),
  mMqttServerPort(mqttServerPort),
  mMqttClient(mqttServerIp, mqttServerPort, mWifiClient)
{
#ifdef WIFI_FIX
   mConnState = 0; // connection state (1-5)
   mWaitCount = 0;
   mBrokerConnectPauseMills = 60*1000;
#else
  // WiFi connection
  mWifiConnected = false;
  mLastWifiConnectionAttemptMillis = 0;
  mLastWifiConnectionSuccessMillis = 0;
#endif

  // MQTT client
  mTopicSubscriptionListSize = 0;
  mMqttConnected = false;
  mLastMqttConnectionMillis = 0;
  mMqttLastWillTopic = 0;
  mMqttLastWillMessage = 0;
  mMqttLastWillRetain = false;
  mMqttCleanSession = true;
  mMqttClient.setCallback([this](char* topic, byte* payload, unsigned int length) {this->mqttMessageReceivedCallback(topic, payload, length);});

  // Web updater
  mUpdateServerAddress = NULL;
  mHttpServer = NULL;
  mHttpUpdater = NULL;

  // other
  mEnableSerialLogs = false;
  mConnectionEstablishedCallback = onConnectionEstablished;
  mShowLegacyConstructorWarning = false;
  mDelayedExecutionListSize = 0;
  mConnectionEstablishedCount = 0;
}

EspMQTTClient::~EspMQTTClient()
{
  if (mHttpServer != NULL)
    delete mHttpServer;
  if (mHttpUpdater != NULL)
    delete mHttpUpdater;
}


// =============== Configuration functions, most of them must be called before the first loop() call ==============

void EspMQTTClient::enableDebuggingMessages(const bool enabled)
{
  mEnableSerialLogs = enabled;
}

#ifdef WIFI_FIX
void EspMQTTClient::configureHTTPWebUpdater()
{
  MDNS.begin(mMqttClientName);
  mHttpUpdater->setup(mHttpServer, mUpdateServerAddress, mUpdateServerUsername, mUpdateServerPassword);
  mHttpServer->begin();
  MDNS.addService("http", "tcp", 80);

  if (mEnableSerialLogs)
    Serial.printf("WEB: Updater ready, open http://%s.local in your browser and login with username '%s' and password '%s'.\n", mMqttClientName, mUpdateServerUsername, mUpdateServerPassword);
}

#endif



void EspMQTTClient::enableHTTPWebUpdater(const char* username, const char* password, const char* address)
{
  if (mHttpServer == NULL)
  {
    mHttpServer = new WebServer(80);
    mHttpUpdater = new ESPHTTPUpdateServer(mEnableSerialLogs);
    mUpdateServerUsername = (char*)username;
    mUpdateServerPassword = (char*)password;
    mUpdateServerAddress = (char*)address;
  }
  else if (mEnableSerialLogs)
    Serial.print("SYS! You can't call enableHTTPWebUpdater() more than once !\n");
}

void EspMQTTClient::enableHTTPWebUpdater(const char* address)
{
  if(mMqttUsername == NULL || mMqttPassword == NULL)
    enableHTTPWebUpdater("", "", address);
  else
    enableHTTPWebUpdater(mMqttUsername, mMqttPassword, address);
}

void EspMQTTClient::enableMQTTPersistence()
{
  mMqttCleanSession = false;
}

void EspMQTTClient::enableLastWillMessage(const char* topic, const char* message, const bool retain)
{
  mMqttLastWillTopic = (char*)topic;
  mMqttLastWillMessage = (char*)message;
  mMqttLastWillRetain = retain;
}


// =============== Public functions =================
#ifdef WIFI_FIX
void EspMQTTClient::loop()
{
  unsigned long currentMillis = millis();
  // start of non-blocking connection setup section
  
  // mConnState values
  // status |   WiFi   |    MQTT
  // -------+----------+------------
  //      0 |   down   |    down
  //      1 | starting |    down
  //      2 |    up    |    down
  //      3 |    up    |  starting
  //      4 |    up    | finalising
  //      5 |    up    |     up

  if ((WiFi.status() != WL_CONNECTED) && (mConnState != 1)) {
    mConnState = 0;
  }
  if ((WiFi.status() == WL_CONNECTED) && !mMqttClient.connected() && (mConnState != 3))  {
    mConnState = 2;
  }
  if ((WiFi.status() == WL_CONNECTED) && mMqttClient.connected() && (mConnState != 5)) {
    mConnState = 4;
  }
  switch (mConnState) {
    case 0:                                                       // MQTT and WiFi down: start WiFi
      if (mEnableSerialLogs)
	Serial.println("WiFi and MQTT down: starting WiFi");
      connectToWifi();
      mConnState = 1;
      break;
    case 1:                                                       // WiFi starting, do nothing here
      if (mEnableSerialLogs) {
	if (mWaitCount == 0) 
	{
	  Serial.printf("WiFi starting, MQTT down, waiting %d ", mWaitCount);
	}
	else if (mWaitCount % 1000 == 0) 
	{
	  Serial.printf("\nWiFi starting, MQTT down, waiting %d ", mWaitCount);
	}
	else if (mWaitCount < 11) 
	{
	  Serial.printf("%d, ", mWaitCount);
	}
	else if (mWaitCount % 50 == 0) 
	{
	  Serial.printf(".", mWaitCount);
	}
      }
      delay(10);
      mWaitCount++;
      break;
    case 2:                                                       // WiFi up, MQTT down: start MQTT
      if (mEnableSerialLogs) {
	Serial.print("\nWiFi up, MQTT down. IP = ");
	Serial.println(WiFi.localIP());
      }
      mWaitCount = 0;
      // Configure web updater
      if (mHttpServer != NULL)
      {
	configureHTTPWebUpdater();
      }
      mTopicSubscriptionListSize = 0;
      mConnState = 3;
      break;
    case 3:                                                       // WiFi up, MQTT starting, do nothing here
      mWaitCount++;
      if (mWaitCount < 3) 
      {
	Serial.println("WiFi up, MQTT starting");
      } 
      if (mWaitCount == 2 || ((mWaitCount % 1000000) == 0)) 
      {
	Serial.print("\nWiFi up, MQTT starting ");
      } 
      else if ((mWaitCount % 20000) == 0)
      {
	Serial.print(".");
      }
      connectToMqttBroker();
      break;
    case 4:                                                       // WiFi up, MQTT up: finish MQTT configuration
      if (mEnableSerialLogs)
      {
	if (mWaitCount < 2) {
	  Serial.println("WiFi up, MQTT up on first try.");
	} 
	else 
	{
	  Serial.println("WiFi up, MQTT up after multiple tries.");
	}
      }
      //mMqttClient.subscribe(output_topic);
      mWaitCount = 0;
      mConnState = 5;
      break;
    case 5:                                                       // Running MQTT
      mMqttClient.loop();
      // Web updater handling
      if (mHttpServer != NULL)
      {
	mHttpServer->handleClient();
	#ifdef ESP8266
	  MDNS.update(); // We need to do this only for ESP8266
	#endif
      }
  }
  // Delayed execution handling
  if (mDelayedExecutionListSize > 0)
  {
    unsigned long currentMillis = millis();

    for(byte i = 0 ; i < mDelayedExecutionListSize ; i++)
    {
      if (mDelayedExecutionList[i].targetMillis <= currentMillis)
      {
	//Serial.printf("Calling # %d, current ms %d, target %d\n", i, mDelayedExecutionList[i].targetMillis, currentMillis);
        (*mDelayedExecutionList[i].callback)();
        for(byte j = i ; j < mDelayedExecutionListSize-1 ; j++)
          mDelayedExecutionList[j] = mDelayedExecutionList[j + 1];
        mDelayedExecutionListSize--;
        i--;
      }
    }
  }

  // Old constructor support warning
  if (mEnableSerialLogs && mShowLegacyConstructorWarning)
  {
    mShowLegacyConstructorWarning = false;
    Serial.print("SYS! You are using a constructor that will be deleted soon, please update your code with the new construction format.\n");
  }
}
#else
void EspMQTTClient::loop()
{
  unsigned long currentMillis = millis();

  if (WiFi.status() == WL_CONNECTED)
  {
    // If we just being connected to wifi
    if (!mWifiConnected)
    {
      if (mEnableSerialLogs)
        Serial.printf("WiFi: Connected, ip : %s\n", WiFi.localIP().toString().c_str());

      mLastWifiConnectionSuccessMillis = millis();
      
      // Config of web updater
      if (mHttpServer != NULL)
      {
	MDNS.begin(mMqttClientName);
        mHttpUpdater->setup(mHttpServer, mUpdateServerAddress, mUpdateServerUsername, mUpdateServerPassword);
        mHttpServer->begin();
        MDNS.addService("http", "tcp", 80);
      
        if (mEnableSerialLogs)
          Serial.printf("WEB: Updater ready, open http://%s.local in your browser and login with username '%s' and password '%s'.\n", mMqttClientName, mUpdateServerUsername, mUpdateServerPassword);
      }
      mWifiConnected = true;
    }
    
    // MQTT handling
    if (mMqttClient.connected())
      mMqttClient.loop();
    else
    {
      if (mMqttConnected)
      {
        if (mEnableSerialLogs)
          Serial.println("MQTT! Lost connection.");
        
        mTopicSubscriptionListSize = 0;
        mMqttConnected = false;
      }
      
      if (currentMillis - mLastMqttConnectionMillis > CONNECTION_RETRY_DELAY || mLastMqttConnectionMillis == 0)
        connectToMqttBroker();
    }
      
    // Web updater handling
    if (mHttpServer != NULL)
    {
      mHttpServer->handleClient();
      #ifdef ESP8266
        MDNS.update(); // We need to do this only for ESP8266
      #endif
    }
      
  }
  else // If we are not connected to wifi
  {
    if (mWifiConnected) 
    {
      if (mEnableSerialLogs)
        Serial.println("WiFi! Lost connection.");
      
      mWifiConnected = false;

      // If we handle wifi, we force disconnection to clear the last connection
      if (mWifiSsid != NULL)
        WiFi.disconnect();
    }
    
    // We retry to connect to the wifi if we handle it and there was no attempt since the last connection lost
    if (mWifiSsid != NULL && (mLastWifiConnectionAttemptMillis == 0 || mLastWifiConnectionSuccessMillis > mLastWifiConnectionAttemptMillis))
      connectToWifi();
  }
  
  // Delayed execution handling
  if (mDelayedExecutionListSize > 0)
  {
    unsigned long currentMillis = millis();

    for(byte i = 0 ; i < mDelayedExecutionListSize ; i++)
    {
      if (mDelayedExecutionList[i].targetMillis <= currentMillis)
      {
        (*mDelayedExecutionList[i].callback)();
        for(byte j = i ; j < mDelayedExecutionListSize-1 ; j++)
          mDelayedExecutionList[j] = mDelayedExecutionList[j + 1];
        mDelayedExecutionListSize--;
        i--;
      }
    }
  }

  // Old constructor support warning
  if (mEnableSerialLogs && mShowLegacyConstructorWarning)
  {
    mShowLegacyConstructorWarning = false;
    Serial.print("SYS! You are using a constructor that will be deleted soon, please update your code with the new construction format.\n");
  }

}
#endif

bool EspMQTTClient::publish(const String &topic, const String &payload, bool retain)
{
  bool success = mMqttClient.publish(topic.c_str(), payload.c_str(), retain);

  if (mEnableSerialLogs) 
  {
    if(success)
      Serial.printf("MQTT << [%s] %s\n", topic.c_str(), payload.c_str());
    else
      Serial.println("MQTT! publish failed, is the message too long ?"); // This can occurs if the message is too long according to the maximum defined in PubsubClient.h
  }

  return success;
}

bool EspMQTTClient::subscribe(const String &topic, MessageReceivedCallback messageReceivedCallback)
{
  // Check the possibility to add a new topic
  if (mTopicSubscriptionListSize >= MAX_TOPIC_SUBSCRIPTION_LIST_SIZE) 
  {
    if (mEnableSerialLogs)
      Serial.println("MQTT! Subscription list is full, ignored.");
    return false;
  }

  // Check the duplicate of the subscription to the topic
  bool found = false;
  for (byte i = 0; i < mTopicSubscriptionListSize && !found; i++)
    found = mTopicSubscriptionList[i].topic.equals(topic);

  if (found) 
  {
    if (mEnableSerialLogs)
      Serial.printf("MQTT! Subscribed to [%s] already, ignored.\n", topic.c_str());
    return false;
  }

  // All checks are passed - do the job
  bool success = mMqttClient.subscribe(topic.c_str());

  if(success)
    mTopicSubscriptionList[mTopicSubscriptionListSize++] = { topic, messageReceivedCallback, NULL };
  
  if (mEnableSerialLogs)
  {
    if(success)
      Serial.printf("MQTT: Subscribed to [%s]\n", topic.c_str());
    else
      Serial.println("MQTT! subscribe failed");
  }

  return success;
}

bool EspMQTTClient::subscribe(const String &topic, MessageReceivedCallbackWithTopic messageReceivedCallback)
{
  if(subscribe(topic, (MessageReceivedCallback)NULL))
  {
    mTopicSubscriptionList[mTopicSubscriptionListSize-1].callbackWithTopic = messageReceivedCallback;
    return true;
  }
  return false;
}

bool EspMQTTClient::unsubscribe(const String &topic)
{
  bool found = false;
  bool success = false;

  for (byte i = 0; i < mTopicSubscriptionListSize; i++)
  {
    if (!found)
    {
      if (mTopicSubscriptionList[i].topic.equals(topic))
      {
        found = true;
        success = mMqttClient.unsubscribe(topic.c_str());

        if (mEnableSerialLogs)
        {
          if(success)
            Serial.printf("MQTT: Unsubscribed from %s\n", topic.c_str());
          else
            Serial.println("MQTT! unsubscribe failed");
        }
      }
    }

    if (found)
    {
      if ((i + 1) < MAX_TOPIC_SUBSCRIPTION_LIST_SIZE)
        mTopicSubscriptionList[i] = mTopicSubscriptionList[i + 1];
    }
  }

  if (found)
    mTopicSubscriptionListSize--;
  else if (mEnableSerialLogs)
    Serial.println("MQTT! Topic cannot be found to unsubscribe, ignored.");

  return success;
}

void EspMQTTClient::executeDelayed(const unsigned long delay, DelayedExecutionCallback callback)
{
  if (mDelayedExecutionListSize < MAX_DELAYED_EXECUTION_LIST_SIZE)
  {
    DelayedExecutionRecord delayedExecutionRecord;
    delayedExecutionRecord.targetMillis = millis() + delay;
    delayedExecutionRecord.callback = callback;
    
    mDelayedExecutionList[mDelayedExecutionListSize] = delayedExecutionRecord;
    mDelayedExecutionListSize++;
    //Serial.printf("Added delayed execution, list size = %d\n", mDelayedExecutionListSize);
  }
  else if (mEnableSerialLogs)
    Serial.printf("SYS! The list of delayed functions is full.\n");
}

#ifdef WIFI_FIX
bool EspMQTTClient::isConnected() // identical to isMqttConnected
{
  //Serial.printf("Connection State = %d\n", mConnState);
  if (mConnState = 5) 
    return true;
  return false;
}

bool EspMQTTClient::isWifiConnected()
{ 
  if (mConnState >= 3) 
    return true;
  return false;
}

bool EspMQTTClient::isMqttConnected()
{
  if (mConnState = 5) 
    return true;
  return false;
}

void EspMQTTClient::brokerConnectPause(unsigned int mills)
{
  mBrokerConnectPauseMills = mills;
}
#endif
// ================== Private functions ====================-

void EspMQTTClient::connectToWifi()
{
  WiFi.mode(WIFI_STA);
  #ifdef ESP32
    WiFi.setHostname(mMqttClientName);
  #else
    WiFi.hostname(mMqttClientName);
  #endif
  WiFi.begin(mWifiSsid, mWifiPassword);

  if (mEnableSerialLogs)
    Serial.printf("\nWiFi: Connecting to %s ... \n", mWifiSsid);
#ifndef WIFI_FIX
  mLastWifiConnectionAttemptMillis = millis();
#endif
}

void EspMQTTClient::connectToMqttBroker()
{
#ifdef WIFI_FIX
  if ((mLastMqttConnectionMillis == 0) || // e.g., first connection
      (mLastMqttConnectionMillis + mBrokerConnectPauseMills < millis())) {
#endif
    if (mEnableSerialLogs) {
      //Serial.printf("Time = %d, Last connect + pause = %d\n", millis(), mLastMqttConnectionMillis + mBrokerConnectPauseMills);
      Serial.printf("MQTT: Connecting to broker @%s ... ", mMqttServerIp);
    }
    if (mMqttClient.connect(mMqttClientName, mMqttUsername, mMqttPassword, mMqttLastWillTopic, 0, mMqttLastWillRetain, mMqttLastWillMessage, mMqttCleanSession))
    {
      mMqttConnected = true;
      
      if (mEnableSerialLogs) 
	Serial.println("ok.");
  
      mConnectionEstablishedCount++;
      (*mConnectionEstablishedCallback)();
    }
    else if (mEnableSerialLogs)
    {
      Serial.print("unable to connect, ");
  
      switch (mMqttClient.state())
      {
	case -4:
	  Serial.println("MQTT_CONNECTION_TIMEOUT");
	  break;
	case -3:
	  Serial.println("MQTT_CONNECTION_LOST");
	  break;
	case -2:
	  Serial.println("MQTT_CONNECT_FAILED");
	  break;
	case -1:
	  Serial.println("MQTT_DISCONNECTED");
	  break;
	case 1:
	  Serial.println("MQTT_CONNECT_BAD_PROTOCOL");
	  break;
	case 2:
	  Serial.println("MQTT_CONNECT_BAD_CLIENT_ID");
	  break;
	case 3:
	  Serial.println("MQTT_CONNECT_UNAVAILABLE");
	  break;
	case 4:
	  Serial.println("MQTT_CONNECT_BAD_CREDENTIALS");
	  break;
	case 5:
	  Serial.println("MQTT_CONNECT_UNAUTHORIZED");
	  break;
      }
    }
    mLastMqttConnectionMillis = millis();
#ifdef WIFI_FIX
  }
#endif
}

/**
 * Matching MQTT topics, handling the eventual presence of a single wildcard character
 *
 * @param topic1 is the topic may contain a wildcard
 * @param topic2 must not contain wildcards
 * @return true on MQTT topic match, false otherwise
 */
bool EspMQTTClient::mqttTopicMatch(const String &topic1, const String &topic2) {
	//Serial.println(String("Comparing: ") + topic1 + " and " + topic2);
	int i = 0;
	if((i = topic1.indexOf('#')) >= 0) {
		//Serial.print("# detected at position "); Serial.println(i);
		String t1a = topic1.substring(0, i);
		String t1b = topic1.substring(i+1);
		//Serial.println(String("t1a: ") + t1a);
		//Serial.println(String("t1b: ") + t1b);
		if((t1a.length() == 0 || topic2.startsWith(t1a))&&
		   (t1b.length() == 0 || topic2.endsWith(t1b)))
		   return true;
	} else if((i= topic1.indexOf('+')) >= 0) {
		//Serial.print("+ detected at position "); Serial.println(i);
		String t1a = topic1.substring(0, i);
		String t1b = topic1.substring(i+1);
		//Serial.println(String("t1a: ") + t1a);
		//Serial.println(String("t1b: ") + t1b);
		if((t1a.length() == 0 || topic2.startsWith(t1a))&&
		   (t1b.length() == 0 || topic2.endsWith(t1b))) {
			if(topic2.substring(t1a.length(), topic2.length()-t1b.length()).indexOf('/') == -1)
				return true;
		}

	} else {
		return topic1.equals(topic2);
	}
	return false;
}

void EspMQTTClient::mqttMessageReceivedCallback(char* topic, byte* payload, unsigned int length)
{
  // Convert the payload into a String
  // First, We ensure that we dont bypass the maximum size of the PubSubClient library buffer that originated the payload
  // This buffer has a maximum length of MQTT_MAX_PACKET_SIZE and the payload begin at "headerSize + topicLength + 1"
  unsigned int strTerminationPos;
  if (strlen(topic) + length + 9 >= MQTT_MAX_PACKET_SIZE)
  {
    strTerminationPos = length - 1;

    if (mEnableSerialLogs)
      Serial.print("MQTT! Your message may be truncated, please change MQTT_MAX_PACKET_SIZE of PubSubClient.h to a higher value.\n");
  }
  else
    strTerminationPos = length;

  // Second, we add the string termination code at the end of the payload and we convert it to a String object
  payload[strTerminationPos] = '\0';
  String payloadStr((char*)payload);
  String topicStr(topic);

  // Logging
  if (mEnableSerialLogs)
    Serial.printf("MQTT >> [%s] %s\n", topic, payloadStr.c_str());

  // Send the message to subscribers
  for (byte i = 0 ; i < mTopicSubscriptionListSize ; i++)
  {
    if (mqttTopicMatch(mTopicSubscriptionList[i].topic, String(topic)))
    {
      if(mTopicSubscriptionList[i].callback != NULL)
        (*mTopicSubscriptionList[i].callback)(payloadStr); // Call the callback
      if(mTopicSubscriptionList[i].callbackWithTopic != NULL)
        (*mTopicSubscriptionList[i].callbackWithTopic)(topicStr, payloadStr); // Call the callback
    }
  }
}

**Comment by abqmichael** [comment]
Hi 

> If you plan to do a pull request, it's better if you finish your cleanup and create a pull request. This way I will easily be able to see changes and comment parts of the code :)

I simply could not find my way through the pull request process. The documentation you suggested assumes that I've done something to prepare my files, like create a git repository of my own or something, but doesn't say what that might be. 

I tried clicking into pull request but it would not let me submit my source. I downloaded the GIT tool, then your repository, checked in a local copy but didn't have permission to upload that to your repository and didn't see how that could help me create something that GIT would recognize/use as a pull request. 

Sorry, but I'm going to have Scotty beam me out of git hell for now. My code is above. Would be happy to email you a copy too.

**Comment by abqmichael** [comment]
Frustration resolved. I found a beginners guide to doing this at https://akrabat.com/the-beginners-guide-to-contributing-to-a-github-project/ and was able to create a fork, push the data up to GitHub and then create a pull request. Kudos to the git developers and documenters for making this relatively simple thing so opaque.

**Comment by abqmichael** [comment]
Here is a state diagram explaining the new connection algorithm.

![EspMqttClient WiFi Connection state diagram V2](https://user-images.githubusercontent.com/47368837/73279598-a06d8800-41aa-11ea-9108-6dfef8307da6.png)

**Comment by plapointe6** [comment]
It will take time to review your changes and test them with an ESP32 and ESP8266, please be patient and thanks for your contribution.

**Comment by abqmichael** [comment]
Thanks for letting me know. I have been using it on 3 devices since posting about it and it’s working great. Not one crash. I monitored one ESP32 via the serial port for several days and found that it would have crashed four times over that period without the changes. My only timing concern would be merging my changes back in if the library was updated without the changes.

Michael J.

**Comment by ssp5zone** [comment]
Not sure if you gave a fix. But seem to be working now. 🥳 

Just tested it on ESP32-WROOM.

> 01:40:22.633 -> WiFi! Lost connection.
> 01:40:22.633 -> 
> 01:40:22.633 -> WiFi: Connecting to Wi5Zone ... 
> 01:40:41.851 -> WiFi: Connecting to Wi5Zone ... 
> 01:41:00.197 -> WiFi: Connecting to Wi5Zone ... 
> 01:41:22.332 -> WiFi: Connected, ip : 192.168.0.100
> 01:41:22.332 -> MQTT: Connecting to broker @io.adafruit.com ... ok.

Cheers! 🍻

**Comment by plapointe6** [comment]
Version 1.9.0 introduce an overhaul of the connection mecanism to make the ESP32 connection a lot more stable. I will close this issue, but do not hesitate to re-open another issue more recent with updated details if there still ESP32 wifi connection issues. Regards.
